File: modules/format/src/main/scala/be/wegenenverkeer/atomium/format/Url.scala
package be.wegenenverkeer.atomium.format

/**
 * A wrapper for an URL that adds a path constructor.
 * @param path url path
 */
case class Url(path: String) {

  /**
   * Creates a new URL by adding the given path Url
   *
   * @param additionalPath the path to be added
   * @return the new URL
   */
  def /(additionalPath: Url) : Url = this / additionalPath.path

  /**
   * Creates a new URL by adding the given path as String
   *
   * @param additionalPath the path to be added
   * @return the new URL
   */
  def /(additionalPath: String): Url = {
    add(additionalPath)
  }

  def /(additionalPath: Long): Url =
    add(additionalPath.toString)

  def /(additionalPath: Int): Url =
    add(additionalPath.toString)

  private def add(additional: String) : Url= {

    val startsWith = additional.startsWith("/")
    val endsWith = path.endsWith("/")

    if(startsWith && endsWith) {
      //one of the slashes must go
      Url(s"$path${additional.drop(1)}")
    } else if( startsWith || endsWith) {
      // only one slash, ok
      Url(s"$path$additional")
    } else {
      // no slashes, add one
      Url(s"$path/$additional")
    }

  }
}

File: modules/core/src/main/java/be/wegenenverkeer/atomium/format/Url.java
package be.wegenenverkeer.atomium.format;

/**
 * Created by Karel Maesen, Geovise BVBA on 30/09/16.
 */
public class Url {

    private final String path;

    public Url(String path) {
        this.path = path;
    }


    public String getPath() {
        return this.path;
    }

    public Url add(Url otherUrl){
        return this.add(otherUrl.getPath());
    }

    public Url add(String additional) {
        boolean startsWith = additional.startsWith("/");
        boolean endsWith = this.path.endsWith("/");

        if (startsWith && endsWith) {
            return new Url(this.path + additional.substring(1));
        } else if (startsWith || endsWith) {
            return new Url(this.path+additional);
        } else {
            return new Url(this.path + "/" + additional);
        }
    }

    public Url add(Integer additionalPath) {
        return add(additionalPath.toString());
    }

    public Url add(Long additionalPath) {
        return add(additionalPath.toString());
    }

}
